wolfCLU/src/crypto/clu_crypto_setup.c

```diff
@@ -0,0 +1,358 @@
+    char     inName[256];       /* name of the in File if not provided */
+
+    char*    name = NULL;       /* string of algorithm, mode, keysize */
+    int      alg;               /* algorithm from name */
+    char*    mode = NULL;       /* mode from name */
+    char*    out  = NULL;       /* default output file name */
+    char*    in = inName;       /* default in data */
+    byte*    pwdKey = NULL;     /* password for generating pwdKey */
+    byte*    key = NULL;        /* user set key NOT PWDBASED */
+    byte*    iv = NULL;         /* iv for initial encryption */
+
+
+    int      size       =   0;  /* keysize from name */
+    int      ret        =   0;  /* return variable */
+    int      block      =   0;  /* block size based on algorithm */
+    int      pwdKeyChk  =   0;  /* if a pwdKey has been provided */
+    int      ivCheck    =   0;  /* if the user sets the IV explicitly */
+    int      keyCheck   =   0;  /* if ivCheck is 1 this should be set also */
+    int      inCheck    =   0;  /* if input has been provided */
+    int      outCheck   =   0;  /* if output has been provided */
+    int      i          =   0;  /* loop counter */
+    int      eCheck     =   0;  /* if user is encrypting data */
+    int      dCheck     =   0;  /* if user is decrypting data */
+    int      inputHex   =   0;  /* if user is encrypting hexidecimal stuff */
+    int      keyType    =   0;  /* tells Decrypt which key it will be using
+                                 * 1 = password based key, 2 = user set key
+                                 */
+    int      printOut   =   0;  /* flag to print out key/iv/salt */
+    int      pbkVersion =   1;
+    const WOLFSSL_EVP_MD* hashType = wolfSSL_EVP_sha256();
+
+    const WOLFSSL_EVP_CIPHER* cphr = NULL;
+    word32   ivSize     =   0;  /* IV if provided should be 2*block */
+    word32   numBits    =   0;  /* number of bits in argument from the user */
+    int      option;
+    int      long_index = 0;
+
+    if (action == 'e')
+        eCheck = 1;
+    if (action == 'd')
+        dCheck = 1;
+
+    ret = wolfCLU_checkForArg("-h", 2, argc, argv);
+    if (ret > 0) {
+        if (eCheck == 1) {
+            wolfCLU_encryptHelp();
+            return 0;
+        } else {
+            wolfCLU_decryptHelp();
+            return 0;
+        }
+    }
+
+    /* gets blocksize, algorithm, mode, and key size from name argument */
+    block = wolfCLU_getAlgo(argc, argv, &alg, &mode, &size);
+    if (block < 0) {
+        printf("unable to find algorithm to use\n");
+        return -1;
+    }
+
+    /* initialize memory buffers */
+    pwdKey = (byte*) XMALLOC(size + block, HEAP_HINT, DYNAMIC_TYPE_TMP_BUFFER);
+    if (pwdKey == NULL)
+        return MEMORY_E;
+    XMEMSET(pwdKey, 0, size + block);
+
+    iv = (byte*) XMALLOC(block, HEAP_HINT, DYNAMIC_TYPE_TMP_BUFFER);
+    if (iv == NULL) {
+        wolfCLU_freeBins(pwdKey, NULL, NULL, NULL, NULL);
```

Comment:
This strikes me as falling into the same category as the "free_stuff" functions. I would just free the memory. Having a wrapper function that does nothing besides checking for null and calling free just seems excessive, to me at least.

Reply:
It will take a bit of time to refactor but I can add fall through on ret checks and free at the end of the function to remove that old wolfCLU_freeBins function.